### DISCUSSION TITLE
App Router / Pass data from children to layout

### DISCUSSION THREAD

**Discussion by r1tsuu** [vercel/next.js#54187] [Help]
Hello.

I have app folder structure:
```
[locale] -> layout.js, post -> [slug] -> page.js 
```
I need to fetch data in the page component and pass that to layout.js?
How could i do this?

My case: Posts have different slugs for different language, component LanguageSwitch is rendered in layout, i need to pass slugs object that i got from the API to LanguageSwitch from post page.

I think i can do this with contexts, but here i probably would need to "reset" context on pages that have the same slug for different languages

Or simpler case: 
Layout has Header component that has backgroundColor property
In page about i need to have header bg red
In page products i need to have header bg blue

Also data is dynamic from API, so i can't do this statically

I could go with contexts here as well (or dom manipulation), but is there something i could do via Next, like pass headerBackgroundColor as prop to layout?

**Comment by jansengunderson** [comment]
We ran into the same issue and are now trying to solve this using contexts. From the [Nextjs docs](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts): 

> Passing data between a parent layout and its children is not possible. However, you can fetch the same data in a route more than once, and React will [automatically dedupe the requests](https://nextjs.org/docs/app/building-your-application/caching#request-memoization) without affecting performance.

**Reply by r1tsuu** [reply]
For my case with LanguageSwitch i didn't found anything better than storing alternate urls in meta tags (generate them with generateMetadata for each language) and in LanguageSwitch just use querySelector. 
Context could work too, but as i described with them you need to reset context state on every page.

```jsx
  const pathname = usePathname(); // next-intl
  const dropdown = useDisclosure();
  const ref = useRef();
  const routerIntl = useRouterIntl(); // next-intl
  const activeLocale = useLocale(); // next-intl

  useOutsideClick(ref, () => {
    if (dropdown.isOpen) dropdown.onClose();
  });

  const handleLanguageChange = (locale) => {
    const linkEl = document.head.querySelector(
      `link[rel="alternate"][hreflang="${locale === 'ua' ? 'uk' : locale}"]`
    );

    const alternatePathname = linkEl
      ? linkEl.getAttribute('href').replace(SITE_URL, '').replace(`/${locale}`, '')
      : null;

    routerIntl.push(alternatePathname ? alternatePathname : pathname, { locale });

    dropdown.onClose();
  };
```

**Comment by S550Stang** [comment]
Re your simpler case: i created a simple project to show you how we can leverage middleware 
```jsx
Layout has Header component that has backgroundColor property
In page about i need to have header bg red
In page products i need to have header bg blue
```
[Middleware](https://github.com/S550Stang/middleware) <----- hope this helps 

if this helps don't forget to upvote this post so other people going through the same issue can benefit from this as well
